### ISSUE TITLE
Resource Group Create SDK should return an async poller - otherwise SDK users who create resources right after creating resource groups may see intermittent resource creation errors

### ISSUE THREAD

**Issue by jake-billings** [Azure/azure-sdk-for-go#21608]
### Our Symptom

We are a team that runs frequent automated tests using the azure-sdk-for-go. Our test job does the following:

1. Create a resource group in ARM
2. Create a resource in that resource group in ARM
3. Delete that resource in ARM
4. Delete that resource group in ARM

Our tests recently started failing intermittently.

Based on our tests, we believe that ARM takes a few seconds for resource group creation to propagate in some cases and that those cases have recently become more common for us.

We have the following logs showing the calls we made to ARM:

| TIMESTAMP | operationName | httpMethod | httpStatusCode |
|-----------|---------------|------------|----------------|
| 2023-09-19T01:49:53.3022402Z | PUT/SUBSCRIPTIONS/RESOURCEGROUPS/ | PUT | -1 |
| 2023-09-19T01:49:53.5115976Z | PUT/SUBSCRIPTIONS/RESOURCEGROUPS/ | PUT | 201 |
| 2023-09-19T01:49:53.5385677Z | PUT/SUBSCRIPTIONS/RESOURCEGROUPS/PROVIDERS/[REDACTED]/[REDACTED]/ | PUT | -1 |
| 2023-09-19T01:49:53.675967Z | PUT/SUBSCRIPTIONS/RESOURCEGROUPS/PROVIDERS/[REDACTED]/[REDACTED]/ | PUT | 404 |
| 2023-09-19T01:49:53.6893615Z | DELETE/SUBSCRIPTIONS/RESOURCEGROUPS/ | DELETE | -1 |
| 2023-09-19T01:49:53.7077024Z | GET/SUBSCRIPTIONS/RESOURCEGROUPS/ | GET | -1 |
| 2023-09-19T01:49:53.7208683Z | GET/SUBSCRIPTIONS/RESOURCEGROUPS/ | GET | 200 |
| 2023-09-19T01:49:53.8012578Z | DELETE/SUBSCRIPTIONS/RESOURCEGROUPS/ | DELETE | 202 |
| 2023-09-19T01:49:57.5268868Z | GET/SUBSCRIPTIONS/RESOURCEGROUPS/ | GET | -1 |
| 2023-09-19T01:49:57.5407255Z | GET/SUBSCRIPTIONS/RESOURCEGROUPS/ | GET | 404 |
| 2023-09-19T01:49:59.8134708Z | 
| 2023-09-19T01:50:54.3947647Z | GET/SUBSCRIPTIONS/RESOURCEGROUPS/ | GET | -1 |
| 2023-09-19T01:50:54.4134174Z | GET/SUBSCRIPTIONS/RESOURCEGROUPS/ | GET | 404 |
| 2023-09-19T01:50:58.6345241Z | GET/SUBSCRIPTIONS/RESOURCEGROUPS/ | GET | -1 |
| 2023-09-19T01:50:58.6453696Z | GET/SUBSCRIPTIONS/RESOURCEGROUPS/ | GET | 404 |
| 2023-09-19T01:54:54.4810864Z | GET/SUBSCRIPTIONS/RESOURCEGROUPS/ | GET | -1 |
| 2023-09-19T01:54:54.4928623Z | GET/SUBSCRIPTIONS/RESOURCEGROUPS/ | GET | 404 |
| 2023-09-19T01:54:57.8525784Z | GET/SUBSCRIPTIONS/RESOURCEGROUPS/ | GET | -1 |
| 2023-09-19T01:54:57.8741165Z | GET/SUBSCRIPTIONS/RESOURCEGROUPS/ | GET | 404 |
| 2023-09-19T01:59:55.1376881Z | GET/SUBSCRIPTIONS/RESOURCEGROUPS/ | GET | -1 |
| 2023-09-19T01:59:55.1507581Z | GET/SUBSCRIPTIONS/RESOURCEGROUPS/ | GET | 404 |
| 2023-09-19T01:59:56.3132961Z | GET/SUBSCRIPTIONS/RESOURCEGROUPS/ | GET | -1 |
| 2023-09-19T01:59:56.3257017Z | GET/SUBSCRIPTIONS/RESOURCEGROUPS/ | GET | 404 |
| 2023-09-19T02:19:59.0603705Z | GET/SUBSCRIPTIONS/RESOURCEGROUPS/ | GET | -1 |
| 2023-09-19T02:19:59.0735594Z | GET/SUBSCRIPTIONS/RESOURCEGROUPS/ | GET | 404 |
| 2023-09-19T02:49:54.5073409Z | GET/SUBSCRIPTIONS/RESOURCEGROUPS/ | GET | -1 |
| 2023-09-19T02:49:54.5207871Z | GET/SUBSCRIPTIONS/RESOURCEGROUPS/ | GET | 404 |

### The Problem in azure-sdk-for-go

Here's the call signature for creation of a resource group in the track 2 azure-sdk-for-go from `azresources.ResourceGroupsClient`:

```go
CreateOrUpdate(
		ctx context.Context,
		resourceGroupName string,
		parameters azresources.ResourceGroup,
		options *azresources.ResourceGroupsClientCreateOrUpdateOptions,
) (response azresources.ResourceGroupsClientCreateOrUpdateResponse, err error)
```

Link: https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/resourcemanager/resources/armresources/resourcegroups_client.go

That type is simply a resource group:

```go
// ResourceGroupsClientCreateOrUpdateResponse contains the response from method ResourceGroupsClient.CreateOrUpdate.
type ResourceGroupsClientCreateOrUpdateResponse struct {
	// Resource group information.
	ResourceGroup
}
```

Link: https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/resourcemanager/resources/armresources/response_types.go

This makes it easy for clients of the sdk to write code like this:

```go
rg, err := resourceGroupClient.CreateOrUpdate(ctx, params.Name, azresources.ResourceGroup{
				Location: to.Ptr(params.Location),
				Name:     to.Ptr(params.Name),
			}, nil)

someOtherResource, err := resourceClient.CreateOrUpdate(ctx, params.Name, rg.Name, // other params)
```

However, this code will have some small chance of failing if ARM encounters the same delay we encountered. Customers of the SDK will go through a difficult and confusing troubleshooting process before either adding their own custom polling or a delay to their code like this:

```go
rg, err := resourceGroupClient.CreateOrUpdate(ctx, params.Name, azresources.ResourceGroup{
				Location: to.Ptr(params.Location),
				Name:     to.Ptr(params.Name),
			}, nil)

time.Sleep(10 * time.Second)

someOtherResource, err := resourceClient.CreateOrUpdate(ctx, params.Name, rg.Name, // other params)
```

We added a custom poller to our code to get around this issue.

Can the resource group client be updated to return an async poller to reflect this delay and to avoid duplicating the troubleshooting we did?

**Comment by raych1** [comment]
@jake-billings , this request needs service side change as currently this API is not an LRO. When it's changed to be an LRO, the generated sdk code would return the async poller.

**Comment by jake-billings** [comment]
I don't think ARM will change the RG implementation here. I think the SDK will need to poll on RGs or users will hit this issue sometimes. The poller might to be a little different from the standard LRO poller for resources.

**Comment by serbrech** [comment]
If this is an issue caused by data propagation (cache or otherwise), then it can occur on all resource types, not just resource groups.

This basically suggests that issuing a GET after the synchronous PUT returned 200 might return 404?

**Comment by jake-billings** [comment]
I have only observed this issue with resource groups. It may or may not happen with other resource types.

I know other resource types that require LRO for creation return a poller. What would that poller do if the operation status returned a 404 once before returning the pending operation status? Would it treat that as part of the polling process or error out?

**Comment by raych1** [comment]
@jake-billings , that's the polling process until the resource get returns 200. It's same behavior across all languages, and that special polling should be taken care of by end user.

**Comment by raych1** [comment]
@jake-billings  all the SDK code are automatically generated, and adding customized code isn't supported by current generation infrastructure.  I talked to other SDK language team, and they haven't received any customer feedback on this issue.  

We will re-evaluate this request if hearing more customer request on this.

**Comment by serbrech** [comment]
Can you get hold of the devs owning. The resourcegroup service to make sure their resource creation is synchronous? We agree that this is a swagger definition issue, if the creation isn't synchronous, and that's not an SDK bug.

**Comment by jake-billings** [comment]
I have reached out to the devs owning resource group. I'll see what they say

**Comment by jake-billings** [comment]
@serbrech I confirmed with ARM team that this behavior is currently expected.

**Comment by raych1** [comment]
> @serbrech I confirmed with ARM team that this behavior is currently expected.

Thanks for your update. I closed this issue.

**Comment by jake-billings** [comment]
@raych1 @serbrech Would customers agree this behavior is expected?

Is there a plan in place to document this - maybe we can provide some example SDK code?

Customers will need to add their own custom polling logic when creating a resource group then creating a resource in that group if they want it to work consistently.

To get it to work, they will need custom code polling 404 responses on the PUT to their new resource. It would appear from the SDK that these responses will not occur or will be retried when they do, but this is not the case.

**Comment by serbrech** [comment]
@jake-billings not at all.
As a matter of fact, AKS Resource Provider was just affected by it, and the QoS of creating AKS cluster was affected by this behavior. to fix it, we have to poll the Resource Group manually after creating it... which should be built-in the SDK.

ARM side fix could be as simple as changing the x-ms option for `final-state-via`, so that the generated poller just issues GET requests on the resource. they wouldn't need to implement the LRO header.

**Comment by serbrech** [comment]
@jake-billings

TL;DR: ensure the calls use the same correlation-id

1. the delay is caused by regional replication
2. it got more visible due to a change in load balancing on incoming calls

The details aren't important, and subject to change.

The more important information is that if the calls provide the **x-ms-correlation-request-id** header, the requests will be routed to the same region, and thus you will not get 404.

Polling the resource group following the PUT will not guarantee that the later operation transit via the same region.

Make sure you pass the same GUID in this header in the calls, and the problem should disappear.

using https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azcore@v1.9.1/policy#WithHTTPHeader

```golang
var cid http.Header
cid.Add("x-ms-correlation-request-id", "8337dcf7-ef97-4a04-8b2a-fdb0e95b387f")
ctx = policy.WithHTTPHeader(ctx, cid)
client.DoSomething(ctx, ...)
```

**Comment by jake-billings** [comment]
Thanks, @serbrech! Will let my team know about this fix!

I wonder if there's any way to pass a correlation id or similar header automatically... 🤔

**Comment by jake-billings** [comment]
@serbrech This fix did not work for us. We continue to see the issue with the correlation id set.

It looks like even with the correlation id being the same, it is possible for traffic to be routed to different regions in quick succession.

| PreciseTimeStamp           | httpMethod | operationName                                                                                          | httpStatusCode | correlationId                            | Location   |
|----------------------------|------------|--------------------------------------------------------------------------------------------------------|----------------|------------------------------------------|------------|
| 2024-02-29T16:23:52.4732221Z | PUT        | PUT/SUBSCRIPTIONS/RESOURCEGROUPS/                                                                      | -1             | 5309fae3-d2d3-49ae-8e59-776c19847c89   | Japan East |
| 2024-02-29T16:23:52.5997332Z | PUT        | PUT/SUBSCRIPTIONS/RESOURCEGROUPS/                                                                      | 201            | 5309fae3-d2d3-49ae-8e59-776c19847c89   | Japan East |
| 2024-02-29T16:23:52.6111353Z | GET        | GET/SUBSCRIPTIONS/RESOURCEGROUPS/                                                                       | -1             | 5309fae3-d2d3-49ae-8e59-776c19847c89   | Japan East |
| 2024-02-29T16:23:52.6316374Z | GET        | GET/SUBSCRIPTIONS/RESOURCEGROUPS/                                                                       | 200            | 5309fae3-d2d3-49ae-8e59-776c19847c89   | Japan East |
| 2024-02-29T16:23:52.6604192Z | PUT        | PUT/SUBSCRIPTIONS/RESOURCEGROUPS/PROVIDERS/MICROSOFT.EXTENDEDLOCATION/CUSTOMLOCATIONS/              | -1             | 5309fae3-d2d3-49ae-8e59-776c19847c89   | Japan West |
| 2024-02-29T16:23:52.777555Z  | PUT        | PUT/SUBSCRIPTIONS/RESOURCEGROUPS/PROVIDERS/MICROSOFT.EXTENDEDLOCATION/CUSTOMLOCATIONS/              | 404            | 5309fae3-d2d3-49ae-8e59-776c19847c89   | Japan West |

**Comment by jake-billings** [comment]
@serbrech @raych1 This behavior is ongoing. Can we please reopen this issue and considering adding a poller here in SDK? It appears that ARM does not currently offer strong consistency, and I believe customers will experience this issue today if they try to create resource groups and resources at the same time.

**Comment by serbrech** [comment]
We also still see it occuring with the correlationID set.
Polling is not helping either since 200 from 1 GET does not guarantee that a subsequent GET will succeed as the data is not consistent depending on where our request lands.
